### DISCUSSION TITLE
Which RSA Key Exchange or/and RSA Signature algorithms do Openssl 1.1.1 use for ciphers AES256-GCM-SHA384 and ECDHE-RSA-CAMELLIA128-SHA256?

### DISCUSSION THREAD

**Discussion by Syllinia** [openssl/openssl#21744] [Q&A] [answered]
Hi Team,

I know that both key exchange and authenticate methods of AES256-GCM-SHA384 is RSA, the authentication method of ECDHE-RSA-CAMELLIA128-SHA256 is RSA.

I'm using 1_1_1_stable branch and would like to figure out which RSA key exchange or/and RSA signature algorithm listed below 1_1_1_stable is using for ciphers AES256-GCM-SHA384 and ECDHE-RSA-CAMELLIA128-SHA256. Thanks in advance.

■RSA Key Exchange
RSA-KEM
RSA-OAEP
RSA-PKCS#1 v1.5
RSA(raw RSA, a raw RSA encryption method that does not perform padding processing)

■RSA Signature
RSA-PSS
RSA-PKCS#1 v1.5
RSA(raw RSA, a raw RSA encryption method that does not perform padding processing)

**Comment by mattcaswell** [comment]
For AES256-GCM-SHA384, RSA key exchange is used. RFC 5264 section 7.4.7.1 says this about it:

>    The RSAES-OAEP encryption scheme defined in [PKCS1] is more secure
   against the Bleichenbacher attack.  However, for maximal
   compatibility with earlier versions of TLS, this specification uses
   the RSAES-PKCS1-v1_5 scheme.  No variants of the Bleichenbacher
   attack are known to exist provided that the above recommendations are
   followed.

In the case of ECDHE-RSA-CAMELLIA128-SHA256, RSA is used for signing only (not key exchange) and an additional ServerKeyExchange message is sent containing the signature. The actual signature scheme used is negotiated separately to the ciphersuite based on the capabilities of the client and server and could be either RSASSA-PSS or RSASSA-PKCS1-v1_5

**Reply by Syllinia** [reply]
Thanks for your reply.
I know there're specifications on RFC but implementation of OSSL is another thing. And my concern is the behavior of 1_1_1_stable.

I dig the related source code and found in rsa_ossl_public_encrypt 1_1_1_stable should support the following RSA encryption scheme:
RSA-OAEP
RSA-PKCS#1 v1.5
RSA(plain RSA, a raw RSA encryption method that does not perform padding processing)

I'm not sure which RSA encryption scheme is used for different cipher suites such as AES256-GCM-SHA384 and ECDHE-RSA-CAMELLIA128-SHA256. As for the RSA signature scheme, I can find the negotiated one in ServerKeyExchange.

**Reply by mattcaswell** [reply, marked as answer]
> I know there're specifications on RFC but implementation of OSSL is another thing.

OpenSSL conforms to the specifications. So it uses PKCS#1 v1.5 for key exchange, and either PSS or PKCS#1 v1.5 for signatures

**Reply by Syllinia** [reply]
I use gdb to validate the key exchange algorithm and find NO_PADDING is actually used.
I tried the below ciphers.
curl -k --ciphers AES256-SHA256 https://10.200.20.133:8443
curl -k --ciphers CAMELLIA256-SHA256 https://xxxxxxx

Below is the backtrace.
(gdb) bt
#0  RSA_private_decrypt (flen=256, 
    from=0x62a000096206 "\036\375\267\327\370\311\177F\371q\247]2\246\023\276\314b\300^\300\273\325\337\037\266(N\316\260\026\254\006\365Wq\240\360\356\231\262\062\\\202\271An\230\210\215r\306\226\363\023o\255\305\365\036\263p\326\214\065\r\327\272S\223\272CPLJt\375_D\"k\023\355\063%\302\346\bSn\344o\027'K?\313\276\272\066F\300\247KKc\311\331\372\n./\227=\214\243\250\r\244\225\325\247q%\362\017z\340\020F\253J\314\373\370", to=0x61100033d000 '\276' <repeats 200 times>..., rsa=0x60f0002728f0, padding=3) at crypto/rsa/rsa_crpt.c:42
#1  0x00007f364e693c9a in tls_process_cke_rsa (pkt=<optimized out>, s=0x61f000072a80) at ssl/statem/statem_srvr.c:3201
#2  tls_process_client_key_exchange (s=0x61f000072a80, pkt=<optimized out>) at ssl/statem/statem_srvr.c:3653
#3  0x00007f364e680dfc in read_state_machine (s=0x61f000072a80) at ssl/statem/statem.c:646
#4  state_machine (s=0x61f000072a80, server=1) at ssl/statem/statem.c:439
#5  0x00007f364e66b224 in SSL_do_handshake (s=0x61f000072a80) at ssl/ssl_lib.c:3836
#6  0x00005563d501bcc4 in ssl_handshake (ssl=0x61f000072a80, wait=0x7f35fba345a8) at ./ssl/ssl_conn.c:255

And it's made by the following commit.
SHA-1: 5b8fa431ae8eb5a18ba913494119e394230d4b70

* Make RSA key exchange code actually constant-time.

Using RSA_PKCS1_PADDING with RSA_private_decrypt is inherently unsafe.
The API requires writing output on success and touching the error queue
on error. Thus, although the padding check itself is constant-time as of
294d1e36c2495ff00e697c9ff622856d3114f14f, and the logic after the
decryption in the SSL code is constant-time as of
adb46dbc6dd7347750df2468c93e8c34bcb93a4b, the API boundary in the middle
still leaks whether the padding check succeeded, giving us our
much-loved Bleichenbacher padding oracle.

Instead, PKCS#1 padding must be handled by the caller which uses
RSA_NO_PADDING, in timing-sensitive code integrated with the
Bleichenbacher mitigation. Removing PKCS#1 padding in constant time is
actually much simpler when the expected length is a constant (and if
it's not a constant, avoiding a padding oracle seems unlikely), so just
do it inline.

Signed-off-by: Kurt Roeckx <kurt@roeckx.be>
Reviewed-by: Rich Salz <rsalz@openssl.org>

GH: #1222

**Comment by kroeckx** [comment]
The padding check and removal happens in the TLS code because it was not possible to do it on a secure way in the crypto code at the time.

**Reply by Syllinia** [reply]
Sorry, I'm a bit confused.

In the Padding section of the following article, it says that plain RSA is vulnerable and there're several attacks against it. And in my gdb process, I found tls_process_cke_rsa actually uses NO_PADDING and removes RSA_PKCS1_PADDING long ago.
https://en.wikipedia.org/wiki/RSA_(cryptosystem)

Can we see commit 5b8fa431ae8eb5a18ba913494119e394230d4b70 is a mitigation for Bleichenbacher padding oracle?
https://medium.com/@c0D3M/bleichenbacher-attack-explained-bc630f88ff25

**Reply by mattcaswell** [reply]
PKCS#1 v1.5 padding is still used. The commit you point to is just moving where the padding is removed from doing it in the underlying RSA implementation to doing it in the TLS implementation, i.e. NO_PADDING is used to get the raw RSA decrypt and then the TLS implementation removes the padding itself.
